File: ios/Classes/SquareInAppPaymentsFlutterPlugin.m
/*
 Copyright 2018 Square Inc.
 
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 
 http://www.apache.org/licenses/LICENSE-2.0
 
 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
*/

#import "SquareInAppPaymentsFlutterPlugin.h"
#import "FSQIPCardEntry.h"
#import "FSQIPApplePay.h"
#import "FSQIPErrorUtilities.h"
#import "FSQIPSecureRemoteCommerce.h"


@interface SquareInAppPaymentsFlutterPlugin ()

@property (strong, readwrite) FSQIPCardEntry *cardEntryModule;
@property (strong, readwrite) FSQIPApplePay *applePayModule;
@property (strong, readwrite) FSQIPSecureRemoteCommerce *secureRemoteCommerceModule;
@end

FlutterMethodChannel *_channel;


@implementation SquareInAppPaymentsFlutterPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar
{
    FlutterMethodChannel *channel = [FlutterMethodChannel
        methodChannelWithName:@"square_in_app_payments"
              binaryMessenger:[registrar messenger]];
    _channel = channel;
    SquareInAppPaymentsFlutterPlugin *instance = [[SquareInAppPaymentsFlutterPlugin alloc] init];
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (instancetype)init
{
    self = [super init];
    if (!self) {
        return nil;
    }
    self.cardEntryModule = [[FSQIPCardEntry alloc] init];
    [self.cardEntryModule initWithMethodChannel:_channel];
    self.applePayModule = [[FSQIPApplePay alloc] init];
    [self.applePayModule initWithMethodChannel:_channel];
    self.secureRemoteCommerceModule = [[FSQIPSecureRemoteCommerce alloc] init];
    [self.secureRemoteCommerceModule initWithMethodChannel:_channel];
    return self;
}

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result
{
    if ([@"setApplicationId" isEqualToString:call.method]) {
        NSString *applicationId = call.arguments[@"applicationId"];
        SQIPInAppPaymentsSDK.squareApplicationID = applicationId;
        result(nil);
    } else if ([@"startCardEntryFlow" isEqualToString:call.method]) {
        BOOL collectPostalCode = [call.arguments[@"collectPostalCode"] boolValue];
        [self.cardEntryModule startCardEntryFlow:result collectPostalCode:collectPostalCode];
    } else if ([@"startGiftCardEntryFlow" isEqualToString:call.method]) {
        [self.cardEntryModule startGiftCardEntryFlow:result];
    } else if ([@"startCardEntryFlowWithBuyerVerification" isEqualToString:call.method]) {
        BOOL collectPostalCode = [call.arguments[@"collectPostalCode"] boolValue];
        NSString *squareLocationId = call.arguments[@"squareLocationId"];
        NSString *buyerActionString = call.arguments[@"buyerAction"];
        NSDictionary *moneyMap = call.arguments[@"money"];
        NSDictionary *contactMap = call.arguments[@"contact"];
        [self.cardEntryModule startCardEntryFlowWithVerification:result
            collectPostalCode:collectPostalCode
            locationId:squareLocationId
            buyerActionString:buyerActionString
            moneyMap:moneyMap
            contactMap:contactMap];
    } else if ([@"completeCardEntry" isEqualToString:call.method]) {
        [self.cardEntryModule completeCardEntry:result];
    } else if ([@"showCardNonceProcessingError" isEqualToString:call.method]) {
        [self.cardEntryModule showCardNonceProcessingError:result errorMessage:call.arguments[@"errorMessage"]];
    } else if ([@"setFormTheme" isEqualToString:call.method]) {
        NSDictionary *theme = call.arguments[@"theme"];
        [self.cardEntryModule setTheme:result theme:theme];
    } else if ([@"initializeApplePay" isEqualToString:call.method]) {
        [self.applePayModule initializeApplePay:result merchantId:call.arguments[@"merchantId"]];
    } else if ([@"canUseApplePay" isEqualToString:call.method]) {
        [self.applePayModule canUseApplePay:result];
    } else if ([@"requestApplePayNonce" isEqualToString:call.method]) {
        NSString *countryCode = call.arguments[@"countryCode"];
        NSString *currencyCode = call.arguments[@"currencyCode"];
        NSString *summaryLabel = call.arguments[@"summaryLabel"];
        NSString *price = call.arguments[@"price"];
        NSString *paymentType = call.arguments[@"paymentType"];
        [self.applePayModule requestApplePayNonce:result
                                      countryCode:countryCode
                                     currencyCode:currencyCode
                                     summaryLabel:summaryLabel
                                            price:price
                                      paymentType:paymentType];
    } else if ([@"completeApplePayAuthorization" isEqualToString:call.method]) {
        BOOL isSuccess = [call.arguments[@"isSuccess"] boolValue];
        NSString *errorMessage = call.arguments[@"errorMessage"];
        [self.applePayModule completeApplePayAuthorization:result isSuccess:isSuccess errorMessage:errorMessage];
    } else if ([@"startBuyerVerificationFlow" isEqualToString:call.method]) {
        NSString *squareLocationId = call.arguments[@"squareLocationId"];
        NSString *buyerActionString = call.arguments[@"buyerAction"];
        NSDictionary *moneyMap = call.arguments[@"money"];
        NSDictionary *contactMap = call.arguments[@"contact"];
        NSString *paymentSourceId = call.arguments[@"paymentSourceId"];
        [self.cardEntryModule startBuyerVerificationFlow:result
                                       buyerActionString:buyerActionString
                                                moneyMap:moneyMap
                                              locationId:squareLocationId
                                              contactMap:contactMap
                                         paymentSourceId:paymentSourceId];
    } else if ([@"startSecureRemoteCommerce" isEqualToString:call.method]) {
        NSNumber *amount = call.arguments[@"amount"];
        [self.secureRemoteCommerceModule startSecureRemoteCommerce:result
                                                            amount:amount.integerValue];
    } else {
        result(FlutterMethodNotImplemented);
    }
}

@end

File: android/src/main/java/sqip/flutter/SquareInAppPaymentsFlutterPlugin.java
/*
 Copyright 2018 Square Inc.

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
*/

package sqip.flutter;

import android.content.Context;
import java.util.HashMap;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.embedding.engine.plugins.activity.ActivityAware;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
import sqip.InAppPaymentsSdk;
import sqip.flutter.internal.CardEntryModule;
import sqip.flutter.internal.GooglePayModule;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

public class SquareInAppPaymentsFlutterPlugin implements MethodCallHandler, FlutterPlugin, ActivityAware {

  private MethodChannel channel;
  private CardEntryModule cardEntryModule;
  private GooglePayModule googlePayModule;
  private Context applicationContext;

  public SquareInAppPaymentsFlutterPlugin() {}

  @Override
  public void onAttachedToEngine(FlutterPluginBinding binding) {
    this.applicationContext = binding.getApplicationContext();
    channel = new MethodChannel(binding.getBinaryMessenger(), "square_in_app_payments");
    // MethodCallHandler is set in onAttachedToActivity to avoid NPE issues
  }

  @Override
  public void onDetachedFromEngine(FlutterPluginBinding binding) {
    cardEntryModule = null;
    googlePayModule = null;
    channel = null;
  }

  @Override
  public void onAttachedToActivity(ActivityPluginBinding activityBinding) {
    channel.setMethodCallHandler(this);
    cardEntryModule = new CardEntryModule(applicationContext, channel);
    googlePayModule = new GooglePayModule(applicationContext, channel);

    googlePayModule.attachActivityResultListener(activityBinding, channel);
    cardEntryModule.attachActivityResultListener(activityBinding, channel);
  }

  @Override
  public void onReattachedToActivityForConfigChanges(ActivityPluginBinding activityBinding) {
    onAttachedToActivity(activityBinding);
  }

  @Override
  public void onDetachedFromActivityForConfigChanges() {
    cardEntryModule = null;
    googlePayModule = null;
    channel = null;
  }

  @Override
  public void onDetachedFromActivity() {
    cardEntryModule = null;
    googlePayModule = null;
    channel = null;
  }

  @Override
  public void onMethodCall(MethodCall call, Result result) {
    switch (call.method) {
      case "setApplicationId":
        String applicationId = call.argument("applicationId");
        InAppPaymentsSdk.INSTANCE.setSquareApplicationId(applicationId);
        result.success(null);
        break;
      case "startCardEntryFlow":
        boolean sce_collectPostalCode = call.argument("collectPostalCode");
        cardEntryModule.startCardEntryFlow(result, sce_collectPostalCode);
        break;
      case "startCardEntryFlowWithBuyerVerification":
        boolean sce_bv_collectPostalCode = call.argument("collectPostalCode");
        String sce_bv_locationId = call.argument("squareLocationId");
        String sce_bv_buyerAction = call.argument("buyerAction");
        HashMap<String, Object> sce_bv_moneyMap = call.argument("money");
        HashMap<String, Object> sce_bv_contactMap = call.argument("contact");
        String sce_bv_paymentSourceId = call.argument("paymentSourceId");

        cardEntryModule.startCardEntryFlowWithBuyerVerification(
          result, 
          sce_bv_collectPostalCode, 
          sce_bv_locationId, 
          sce_bv_buyerAction, 
          sce_bv_moneyMap, 
          sce_bv_contactMap,
          sce_bv_paymentSourceId);
        break;
      case "startGiftCardEntryFlow":
        cardEntryModule.startGiftCardEntryFlow(result);
        break;
      case "startGiftCardEntryFlowWithBuyerVerification":
        String sge_bv_locationId = call.argument("squareLocationId");
        String sge_bv_buyerAction = call.argument("buyerAction");
        HashMap<String, Object> sge_bv_moneyMap = call.argument("money");
        HashMap<String, Object> sge_bv_contactMap = call.argument("contact");
        String sge_bv_paymentSourceId = call.argument("paymentSourceId");
        cardEntryModule.startGiftCardEntryFlowWithBuyerVerification(
          result,
          sge_bv_locationId,
          sge_bv_buyerAction,
          sge_bv_moneyMap,
          sge_bv_contactMap,
          sge_bv_paymentSourceId);
        break;
      case "completeCardEntry":
        cardEntryModule.completeCardEntry(result);
        break;
      case "showCardNonceProcessingError":
        String errorMessage = call.argument("errorMessage");
        cardEntryModule.showCardNonceProcessingError(result, errorMessage);
        break;
      case "initializeGooglePay":
        String igp_squareLocationId = call.argument("squareLocationId");
        int igp_environment = call.argument("environment");
        googlePayModule.initializeGooglePay(igp_squareLocationId, igp_environment);
        result.success(null);
        break;
      case "canUseGooglePay":
        googlePayModule.canUseGooglePay(result);
        break;
      case "requestGooglePayNonce":
        String rgp_price = call.argument("price");
        String rgp_currencyCode = call.argument("currencyCode");
        int rgp_priceStatus = call.argument("priceStatus");
        googlePayModule.requestGooglePayNonce(result, rgp_price, rgp_currencyCode, rgp_priceStatus);
        break;
      case "requestGooglePayNonceWithBuyerVerification":
        String rgp_bv_locationId = call.argument("squareLocationId");
        String rgp_bv_buyerAction = call.argument("buyerAction");
        HashMap<String, Object> rgp_bv_moneyMap = call.argument("money");
        HashMap<String, Object> rgp_bv_contactMap = call.argument("contact");
        String rgp_bv_paymentSourceId = call.argument("paymentSourceId");
        String rgp_bv_price = call.argument("price");
        String rgp_bv_currencyCode = call.argument("currencyCode");
        int rgp_bv_priceStatus = call.argument("priceStatus");

        googlePayModule.requestGooglePayNonceWithBuyerVerification(
          result,
          rgp_bv_buyerAction,
          rgp_bv_moneyMap,
          rgp_bv_locationId,
          rgp_bv_contactMap,
          rgp_bv_paymentSourceId,
          rgp_bv_price,
          rgp_bv_currencyCode,
          rgp_bv_priceStatus);
        break;
      case "startBuyerVerificationFlow":
        String bv_locationId = call.argument("squareLocationId");
        String bv_buyerAction = call.argument("buyerAction");
        HashMap<String, Object> bv_moneyMap = call.argument("money");
        HashMap<String, Object> bv_contactMap = call.argument("contact");
        String bv_paymentSourceId = call.argument("paymentSourceId");

        cardEntryModule.startBuyerVerificationFlow(
          result, 
          bv_buyerAction, 
          bv_moneyMap, 
          bv_locationId, 
          bv_contactMap, 
          bv_paymentSourceId);
        break;
      case "startSecureRemoteCommerce":
        int amount = call.argument("amount");
        // Not yet implemented
        result.notImplemented();
        break;
      case "startSecureRemoteCommerceWithBuyerVerification":
        // Not yet implemented
        result.notImplemented();
        break;
      default:
        result.notImplemented();
        break;
    }
  }
}
